dask/array/core.py

```diff
@@ -2062,7 +2062,7 @@ def asarray(array):
     """
     if not isinstance(array, Array):
         name = 'asarray-' + tokenize(array)
-        if not hasattr(array, 'shape'):
+        if isinstance(getattr(array, 'shape', None), Iterable):
```

Comment:
Yes. The point was that `h5py.Dataset` defined a `.shape` attribute, which indicates that it was already an "array-like" object. So it seemed like the right thing is probably to not eagerly evaluate it -- that was the entire point of this check.

I'm not sure we actually have a test for this functionality, though, so it's totally expected that it broke.